crates/calibration/src/residuals.rs

```diff
@@ -0,0 +1,107 @@
+use types::{CameraPosition, FieldDimensions};
+
+use crate::{corrections::Corrections, lines::LinesError, measurement::Measurement};
+
+pub type Residual = Vector<f32, Dyn, ResidualStorage>;
+pub type ResidualStorage = Owned<f32, Dyn>;
+
+pub fn calculate_residuals_from_parameters(
+    parameters: &Corrections,
+    measurements: &[Measurement],
+    field_dimensions: &FieldDimensions,
+) -> Option<Residual> {
+    measurements
+        .iter()
+        .fold(
+            Some(Vec::new()),
+            |residuals: Option<Vec<f32>>, measurements| match residuals {
+                Some(mut residuals) => {
+                    residuals.extend::<Vec<f32>>(
+                        Residuals::calculate_from(parameters, measurements, field_dimensions)
+                            .ok()?
+                            .into(),
+                    );
+                    Some(residuals)
+                }
+                None => None,
+            },
+        )
+        .map(DVector::from_vec)
+}
+
+pub struct Residuals {
+    pub border_to_connecting_angle: f32,
+    pub connecting_to_goal_box_angle: f32,
+    pub distance_between_parallel_line_start_points: f32,
+    pub distance_between_parallel_line_center_points: f32,
+    pub distance_between_parallel_line_end_points: f32,
+}
+
+impl Residuals {
+    pub fn calculate_from(
+        parameters: &Corrections,
+        measurement: &Measurement,
+        field_dimensions: &FieldDimensions,
+    ) -> Result<Self, ResidualsError> {
+        let corrected = measurement.matrix.to_corrected(
+            parameters.correction_in_robot,
+            match measurement.position {
+                CameraPosition::Top => parameters.correction_in_camera_top,
+                CameraPosition::Bottom => parameters.correction_in_camera_bottom,
+            },
+        );
+
+        let projected_lines = measurement
+            .lines
+            .to_projected(&corrected)
+            .map_err(ResidualsError::NotProjected)?;
+
+        let border_to_connecting_angle = projected_lines
+            .border_line
+            .angle(projected_lines.connecting_line);
+        let connecting_to_goal_box_angle = projected_lines
+            .border_line
+            .angle(projected_lines.connecting_line);
+        let distance_between_parallel_line_start_points = projected_lines
+            .border_line
+            .distance_to_point(projected_lines.goal_box_line.0);
+        let distance_between_parallel_line_center_points = projected_lines
+            .border_line
+            .distance_to_point(projected_lines.goal_box_line.center());
+        let distance_between_parallel_line_end_points = projected_lines
+            .border_line
+            .distance_to_point(projected_lines.goal_box_line.1);
+
+        Ok(Residuals {
+            border_to_connecting_angle: border_to_connecting_angle - FRAC_PI_2,
+            connecting_to_goal_box_angle: connecting_to_goal_box_angle - FRAC_PI_2,
```

Comment:
Are you sure you know what you are doing?
Manually subtracting angles from each other is dangerous, although I couldn't find any obvious issues in this case.

Reply:
Iirc, I'm using the new angle method that always returns a positive angle between [0,180] degrees. Subtracting 90 degrees transforms the interval into the expected one and gives us a sign for the gradient. I think this should work. 😅